pkg/kp/rollstore/consul_store.go

```diff
@@ -0,0 +1,555 @@
+	if err != nil {
+		return consulutil.NewKVError("delete", key, err)
+	}
+	return nil
+}
+
+func (s consulStore) Lock(id roll_fields.ID, session string) (bool, error) {
+	key, err := RollLockPath(id)
+	if err != nil {
+		return false, err
+	}
+
+	success, _, err := s.kv.Acquire(&api.KVPair{
+		Key:     key,
+		Value:   []byte(session),
+		Session: session,
+	}, nil)
+	if err != nil {
+		return false, consulutil.NewKVError("acquire", key, err)
+	}
+	return success, nil
+}
+
+func (s consulStore) Watch(quit <-chan struct{}) (<-chan []roll_fields.Update, <-chan error) {
+	outCh := make(chan []roll_fields.Update)
+	errCh := make(chan error)
+	inCh := make(chan api.KVPairs)
+
+	go consulutil.WatchPrefix(rollTree+"/", s.kv, inCh, quit, errCh)
+
+	go func() {
+		defer close(outCh)
+		defer close(errCh)
+
+		for listed := range inCh {
+			out := make([]roll_fields.Update, 0, len(listed))
+			for _, kvp := range listed {
+				var next roll_fields.Update
+				if err := json.Unmarshal(kvp.Value, &next); err != nil {
+					select {
+					case errCh <- err:
+					case <-quit:
+						// stop processing this kvp list; inCh should be closed
+						// in a moment
+						break
+					}
+				} else {
+					out = append(out, next)
+				}
+			}
+			select {
+			case outCh <- out:
+			case <-quit:
+			}
+		}
+	}()
+
+	return outCh, errCh
+}
+
+func RollPath(id roll_fields.ID) (string, error) {
+	if id == "" {
+		return "", util.Errorf("id not specified when computing roll path")
+	}
+	return path.Join(rollTree, string(id)), nil
+}
+
+// Roll paths are computed using the id of the new replication controller
+func RollLockPath(id roll_fields.ID) (string, error) {
+	subRollPath, err := RollPath(id)
+	if err != nil {
+		return "", err
+	}
+
+	return path.Join(kp.LOCK_TREE, subRollPath), nil
+}
+
+func kvpToRU(kvp *api.KVPair) (roll_fields.Update, error) {
+	ru := roll_fields.Update{}
+	err := json.Unmarshal(kvp.Value, &ru)
+	if err != nil {
+		return ru, util.Errorf("Unable to unmarshal value as rolling update: %s", err)
+	}
+	return ru, nil
+}
+
+// Attempts to create a rolling update. Checks sessionErrCh for session renewal
+// errors just before actually doing the creation to minimize the likelihood of
+// race conditions resulting in conflicting RUs
+func (s consulStore) attemptRUCreation(u roll_fields.Update, sessionErrCh chan error) (roll_fields.Update, error) {
+	b, err := json.Marshal(u)
+	if err != nil {
+		return u, err
+	}
+
+	key, err := RollPath(roll_fields.ID(u.NewRC))
+	if err != nil {
+		return u, err
+	}
+
+	// Confirm that our lock session is still valid, and then create the
+	// rolling update. If session isn't valid, delete the newRC we just
+	// created
+	select {
+	case err := <-sessionErrCh:
+		if err == nil {
+			err = util.Errorf("Cannot create ru because session was destroyed")
+		}
+		return u, err
+	default:
+		success, _, err := s.kv.CAS(&api.KVPair{
+			Key:   key,
+			Value: b,
+			// it must not already exist
+			ModifyIndex: 0,
```

Comment:
By virtue of being CAS, it should be clear that something else can't have created it already (hopefully that's clear to others at least)